src/settings/toml/deploy_target.rs

```diff
@@ -0,0 +1,114 @@
+use crate::settings::toml::Route;
+
+#[derive(Debug)]
+pub struct RouteConfig {
+    pub workers_dev: Option<bool>,
+    pub route: Option<String>,
+    pub routes: Option<Vec<String>>,
+    pub zone_id: Option<String>,
+}
+
+impl RouteConfig {
+    fn has_conflicting_targets(&self) -> bool {
+        if self.workers_dev.unwrap_or_default() {
+            if let Some(pattern) = &self.route {
+                !pattern.is_empty() || self.routes.is_some() // this is all so messy because of deserializer
+            } else {
+                self.routes.is_some()
+            }
+        } else {
+            if let Some(pattern) = &self.route {
+                !pattern.is_empty() && self.routes.is_some()
+            } else {
+                false
+            }
+        }
+    }
+
+    pub fn is_zoneless(&self) -> bool {
+        self.workers_dev.unwrap_or_default() && !self.has_conflicting_targets()
+    }
+
+    pub fn is_complete_zoned(&self) -> bool {
+        (self.route.is_some() || self.routes.is_some()) && self.zone_id.is_some()
+    }
+
+    // zone id is another weird one where `Some("")` is treated the same as `None`
+    pub fn missing_zone_id(&self) -> bool {
+        let result = !self.workers_dev.unwrap_or_default()
+            && (self.route.is_some() || self.routes.is_some());
+        if let Some(zone_id) = &self.zone_id {
+            result && zone_id.is_empty()
+        } else {
+            result
+        }
+    }
+}
+
+impl DeployTarget {
+    pub fn build(script: &str, route_config: &RouteConfig) -> Result<DeployTarget, failure::Error> {
+        if route_config.is_zoneless() {
+            Ok(DeployTarget::Zoneless)
+        } else {
+            // zone_id is required
+            let zone_id = route_config.zone_id.as_ref().unwrap();
```

Comment:
is it possible to make this `&route_config.zone_id` instead, or to replace the unwrap with a `?`

Reply:
nope. its like a &Option<String> and i can't do any of that. i do have it on my list to simplify some of this in a short refactor before i release it though, because some of these intermediary steps are likely unnecessary now that things have been ironed out.